### PR TITLE
fix: improve sections

### LINKED ISSUE #996: Enhanced catalog/profile/guidance management
## Issue description / feature objectives

**Objective** - allow an organisation to devolve ownership for the catalog, profile, guidance and SSP to different individuals or groups

This would be done via separate repositories for catalog, profile, guidance, and SSP (or other implementation documentation), storing the content at each level as markdown and using trestle to transform that into markdown into OSCAL json, to pass to the next level in the chain as a pull request to update the content. 

The flow of content is as follows:
Screenshot 2022-01-20 at 12 43 13

All controls in the catalog flow to the profile as oscal, where a subset of controls is selected and parameters are assigned, to create a profile. The profile then flows to consumers as Oscar, such as guidance and documentation. When guidance (how to meet a control) is updated, the prose is sent to the documentation (how a control is met) as an issue that the guidance for control or group of controls has been updated, and therefore the documentation for those controls must be reviewed to ensure validity.

_For example, a catalog owner updates a control in the catalog:_
1. catalog owner edits markdown of catalog, creates pull request to catalog with proposed changes.
2. catalog approver reviews and approves the changes to the catalog
3. trestle CI detects merged changes to markdown, updates OSCAL json catalog with changes to markdown and commits back to the catalog repo.
4. trestle CI creates a pull request(s) on configured consumers of the catalog (i.e. the profiles repo) requesting the owner of the profiles updates the catalog in the profiles repo.

I think the following functionality is required to enable this:
[ ] - Editable parameters defined in a control in the profile markdown for that control, which are populated into the json profile. For example a list of parameters in AC-1, e.g. ac-1_prm_1, ac-1_prm_2, ac-1_prm_3 in the yaml header for the control, set to the default for that parameter from the catalog, which the profile author can specify, and are then populated into the json profile when `profile-assemble` is run. 
[ ] - Validation that parameters in the profile match the catalog - for example if a control is modified to add an additional parameter, trestle profile validate should populate the profile markdown with the default values from the catalog, and print that new parameters are specified and should be populated by the profile author (which can then be turned into an issue on the profile by the CI/CD). The workflows for create new control in catalog and delete control in catalog will require determination. This is probably part of a wider problem about 'how can we use trestle to validate content and plug in to a CI/CD.
[ ] - The ability to separate guidance from being part of the profile into a separate repo.

### LINKED ISSUE #1012: Define and create new tool to generate only guidance: guidance-generate
## Issue description / feature objectives
Users may need to handle guidance separately from the profile
A new tool is needed to do a form of guidance-generate, to split guidance off from markdown

## Completion Criteria
- [ ] Define how the tool operates and what its format is
- [ ] Implement the tool

### LINKED ISSUE #1029: SSP-assemble should retain details in SystemCharacteristics
## Issue description / feature objectives
Make sure any static content in SystemCharacteristics is retained after ssp-assemble

### LINKED ISSUE #1031: Remove checks for existence of things like args.foo
## Issue description / feature objectives
Several places in the code check for existence of args.foo, but it should always be present (possibly None) if created by an argument parser.  But some tests simply create a namespace and fill it in, in which case all args should be included.
## Completion Criteria
remove all checks for existence of args, e.g. args.foo and update tests as needed.

### PR DESCRIPTION
**Pull request by fsuits**
## Types of changes

- [ ] Hot fix (emergency fix and release)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation (change which affects the documentation site)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Release (`develop` -> `main`)

## Quality assurance (all should be covered).

- [x] My code follows the code style of this project.
- [ ] Documentation for my change is up to date?
- [x] My PR meets testing requirements.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Key links:
Improve handling of sections and required sections
Added required-sections and allowed-sections
Error if missing required sections or have non-allowed section during profile-assemble

closes #1031 
closes #1029 
closes #1012 
closes #996 

- [Sonar coverage](https://sonarcloud.io/dashboard?id=compliance-trestle)

# Before you merge

- Ensure it is a 'squash commit' if not a release.
- Ensure CI is currently passing
- Check sonar. If you are working for a fork a maintainer will reach out, if required.

### CODE REVIEW

File: trestle/core/commands/author/profile.py [hunk 1]
```diff
@@ -86,17 +86,18 @@ def _init_arguments(self) -> None:
...
         )
         self.add_argument('-s', '--sections', help=const.HELP_SECTIONS, required=False, type=str)
+        self.add_argument('-rs', '--required-sections', help=const.HELP_REQUIRED_SECTIONS, required=False, type=str)
 
     def _run(self, args: argparse.Namespace) -> int:
         try:
             log.set_log_level_from_args(args)
-            trestle_root = args.trestle_root
+            trestle_root = pathlib.Path(args.trestle_root)
             if not file_utils.is_directory_name_allowed(args.output):
```
Comment (vikas-agarwal76) [inline trestle/core/commands/author/profile.py:95]: I think trestle_root is already of type Path in all commands.

File: trestle/core/commands/author/profile.py [hunk 2]
```diff
@@ -279,9 +286,10 @@ def assemble_profile(
             md_name: The name of the directory containing the markdown control files for the profile
             new_profile_name: The name of the new json profile.  It can be the same as original to overwrite
             set_parameters: Use the parameters in the yaml header to specify values for setparameters in the profile
-            regenerate: whether to regenerate the uuid's in the profile
-            version: version for the assembled profile
-            required_sections: list of required sections in the assembled profile as comma-separated short names
+            regenerate: Whether to regenerate the uuid's in the profile
+            version: Optional version for the assembled profile
+            required_sections: Optional List of required sections in assembled profile as comma-separated short names
+            allowed_sections: Optional list of section short names that are allowed
```
Comment (vikas-agarwal76) [inline trestle/core/commands/author/profile.py:293]: Is this better -
allowed_sections: Optional list of sections that are allowed in assembled profile as comma separated short names

**Review by vikas-agarwal76 — CHANGES_REQUESTED** [review]
Just some minor comments.

File: trestle/core/commands/author/profile.py (hunk 1 above)
Comment (fsuits) [reply trestle/core/commands/author/profile.py:95]: Correct.  I have changed it but I give it typing to indicate it is type pathlib.Path

File: trestle/core/commands/author/profile.py (hunk 2 above)
Comment (fsuits) [reply trestle/core/commands/author/profile.py:293]: Yes that is better.  Changed.

**Review by vikas-agarwal76 — APPROVED** [review]
LGTM

**Pull request merged**